mrmustard/lab_dev/circuit_components.py

```diff
@@ -463,31 +464,48 @@ def _matmul_indices(self, other: CircuitComponent) -> tuple[tuple[int, ...], tup
         idx_zconj += other.wires.ket.input[ket_modes].indices
         return idx_z, idx_zconj
 
-    def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
+    def __matmul__(self, other: CircuitComponent | Scalar) -> CircuitComponent:
         r"""
         Contracts ``self`` and ``other`` without adding adjoints.
         It allows for a more custom way of contracting components.
         """
+        if isinstance(other, (numbers.Number, np.ndarray)):
+            return self * other
         wires_result, perm = self.wires @ other.wires
         idx_z, idx_zconj = self._matmul_indices(other)
         rep = self.representation[idx_z] @ other.representation[idx_zconj]
         rep = rep.reorder(perm) if perm else rep
         return CircuitComponent._from_attributes(rep, wires_result, None)
 
-    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+    def __rmatmul__(self, other: Scalar) -> CircuitComponent:
+        r"""
+        Multiplies a scalar with a circuit component when written as ``scalar @ component``.
+        """
+        return self * other
+
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent | np.ndarray:
         r"""
         Contracts ``self`` and ``other`` (output of self going into input of other).
-        It adds the adjoints when they are missing. An error is raised if these
-        cannot be deduced from the wires of the components. For example this
-        allows ``Ket``s to be right-shifted into ``Channel``s and automatically
-        the result is a ``DM``:
+        It adds the adjoints when they are missing (e.g. if ``self`` is a Ket and
+        ``other`` is a Channel). An error is raised if these cannot be deduced from
+        the wires of the components. For example this allows ``Ket``s to be right-shifted
+        into ``Channel``s and automatically the result is a ``DM``. If the result has
+        no wires left, it returns the (batched) scalar value of the representation.
```

Comment:
```suggestion
        no wires left, it returns the (batched) scalar value of the representation.
        Note that a ``CircuitComponent`` is allowed to right-shift into scalars because the scalar part may result from an automated contraction subroutine that involves several components).
```